pystac/extensions/datacube.py

```diff
@@ -30,33 +41,65 @@
 DIM_REF_SYS_PROP = "reference_system"
 DIM_UNIT_PROP = "unit"
 
+U = TypeVar("U")
 
-class Dimension(ABC):
-    properties: Dict[str, Any]
-
-    def __init__(self, properties: Dict[str, Any]) -> None:
-        self.properties = properties
 
+class HasProperties(Protocol[U]):
     @property
-    def dim_type(self) -> str:
-        return get_required(
-            self.properties.get(DIM_TYPE_PROP), "cube:dimension", DIM_TYPE_PROP
-        )
+    def properties(self) -> Dict[str, U]:
+        """A dictionary with the properties for the object."""
+
+
+class _Property(Generic[U]):
+    """
+    A descriptor managing the properties of an object.
+
+    Args:
+        key : The key within the managed object's ``properties`` dictionary to use.
+        obj : The key this object is contained with (e.g. ``"cube:variables"``)
+        required : Whether the property is required.
+        clear_if_none : Whether setting a value of `None` clears the value from
+            the ``properties`` dictionary.
+    """
+
+    def __init__(
+        self,
+        key: str,
+        obj: Union[str, Any] = None,
+        required: bool = False,
+        clear_if_none: Optional[bool] = None,
+    ):
+        self.key = key
+        self.obj = obj
+        self.required = required
+        if clear_if_none is None:
+            clear_if_none = not required
+        self.clear_if_none = clear_if_none
+
+    def __get__(self, instance: HasProperties[U], owner: Any = None) -> Optional[U]:
```

Comment:
LMK if the type signatures are confusing (I think they are).

But the usage below will hopefully clear things up. `U` is the type of the actual object being returned, like `str` for description, or `List[float]` for extent.

I see now that we can (maybe) tighten this up a bit. We know that if the property is `required` then return type is just `U` rather than `Optional[U]`. I'll see if mypy can handle that.